crates/sui/src/upgrade_compatibility.rs

```diff
@@ -895,6 +896,36 @@ fn function_lost_public(
     Ok(diags)
 }
 
+fn format_param(
+    param: &Type,
+    type_params: Vec<SourceName>,
+    secondary: &mut Vec<(Loc, String)>,
+) -> Result<String, Error> {
+    Ok(match param {
+        Type::TypeParameter(t) => {
+            let type_param = type_params
+                .get(*t as usize)
+                .context("Unable to get type param location")?;
+
+            secondary.push((
+                type_param.1,
+                format!("Type parameter '{}' is defined here", &type_param.0),
+            ));
+            type_param.0.to_string()
+        }
+        Type::Vector(t) => {
+            format!("vector<{}>", format_param(t, type_params, secondary)?)
+        }
+        Type::MutableReference(t) => {
+            format!("&mut {}", format_param(t, type_params, secondary)?)
+        }
+        Type::Reference(t) => {
+            format!("&{}", format_param(t, type_params, secondary)?)
+        }
+        _ => format!("{}", param),
```

Comment:
Note that this does not handle generic structs and enums (if they contain type parameters, you won't find them and replace them with their type param name). Sorry in advance if I'm pointing something out that you addressed in some later PR, I know this is something we talked about on Wednesday!